text/trace/0170-sampling-probability.md

```diff
@@ -580,92 +586,268 @@ The OTEP 168 proposal for propagation limits head sampling
+probability.  When the probability sampler selects the span, its
+adjusted count will be used.  When the probability sampler does not
+select a span, zero adjusted count will be used.
+
+The use of zero adjusted count allows recording spans that an unbiased
+probability sampler did not select, allowing those spans to be
+received at the backend without introducing statistical bias.
+
+#### Composition rules summary
+
+To create a composite Sampler, first express the result of each
+Sampler in terms of the p-value and `sampled` flag.  Note that
+p-values fall into three categories:
+
+1. Unknown p-value indicates unknown adjusted count
+2. Known non-zero p-value (in the range `[0,62]`) indicates known non-zero adjusted count
+3. Known zero p-value (`p=63`) indicates known zero adjusted count
+
+While non-probability samplers always return unknown `p` and may set
+`sampled=true` or `sampled=false`, a probability sampler is restricted
+to returning either `p∈[0,62]` with `sampled=true` or to returning
+`p=63` with `sampled=false`.  No individual sampler can return `p=63`
+with `sampled=true`, but this condition MAY result from composition of
+`p=63` and unknown `p`.
+
+A composite sampler can be computed using the table below, as follows.
+Although unknown `p` is never encoded in `tracestate`, for the purpose
+of composition we assign unknowns `p=64`, which is 1 beyond the range
+of the 6-bit that represent known p-values.  The assignment of `p=64`
+simplifies the formulas below .
+
+By folowing these simple rules, any numher of consistent probability
+samplers and non-probability samplers can be combined.  Starting with
+`p=64` representing unknown and `sampled=false`, update the composite
+p-value to the minimum value of the prior composite p-value and the
+individual sampler p-value.
 
 ```
-  // Log-head-adjusted count is the logarithm of adjusted count for
-  // this span as calculated at the head, offset by +1, with the
-  // following recognized values.
-  //
-  // 0: The zero value represents an UNKNOWN adjusted count.
-  //    Consumers of these Spans cannot cannot compute span metrics.
-  //
-  // 1: An adjusted count of 1.
-  //
-  // 2-62: Values 2 through 62 represent an adjusted count of 2^(Value-1)
-  //
-  // 63: Value 63 represents an adjusted count of zero.
-  //
-  // Values greater than 64 are unrecognized.
-  uint32 log_head_adjusted_count = <next_tag>;
+p<sub>out</sub> = min(p<sub>in</sub>, p<sub>sampler</sub)
+sampled<sub>out</sub> = logicalOR(sampled<sub>in</sub>, sampled<sub>sampler</sub)
 ```
 
+The composite sampler is always the logical-OR of the individual
+samplers.  For p-value, this has two effects:
+
+1. When combining two consistent probability samplers, the
+less-selective Sampler's adjusted count is taken.
+2. When combining a consistent probability sampler and a
+non-probability sampler, this has the effect of changing unknown
+adjusted count into known adjusted count.
+
```

Comment:
I might be missing something here. Let's consider combining two samplers, one a consistent probability sampler configured with P=10 and a non-probability sampler which selects every 10th span for sampling. We might end up with
Pout = min(10,64) = 10, and
SAMPLEDout=logicalOR(false, true) = true.
This would entail having approximately every 10th span sampled and reported with "adjusted count" of 1024.
What did I miss?

Reply:
Pout will be set to 10 when the span is sampled, 1 out of 1024 times.
Pout will be set to 63 for zero adjusted count by the probability sampler 1023 out of 1024 times.

For every 1024 spans, you expect to record one span with Pout=10 and a bunch (102.3) of Pout=63 spans, which count for nothing but can still be useful. I would say that generally it makes sense to combine probability Samplers with others that are less selective, not more selective.

Reply:
Yes, this is what I'd expect to see. But this is not what I'm getting when going step-by-step over the described algorithm. I think the part
p<sub>out</sub> = min(p<sub>in</sub>, p<sub>sampler</sub)
needs to be under condition
if (sampled<sub>sampler</sub>)

Reply:
The value `p=63` is used by a probability sampler when `!sampled`, that's why this works. I'm trying to cover this on lines 795-800.